src/Common/src/System/Net/Http/TlsCertificateExtensions.cs

```diff
@@ -0,0 +1,134 @@
+                            if (string.Equals(ClientCertificateOid, oid.Value))
+                            {
+                                isClientAuth = true;
+                                break;
+                            }
+                        }
+                    }
+                }
+
+                /// Check if the extension is a key usage extension.
+                /// No point going over it if we have already established that out cert has digital signature
+                if (!foundKeyUsages)
+                {
+                    X509KeyUsageExtension usageExt = extension as X509KeyUsageExtension;
+                    if (usageExt != null)
+                    {
+                        foundKeyUsages = true;
+                        isDigitalSignature = (usageExt.KeyUsages & X509KeyUsageFlags.DigitalSignature) != 0;
+                    }
+                }
+
+                if (foundKeyUsages && foundEku)
+                {
+                    break;
+                }
+            }
+
+            return isClientAuth && isDigitalSignature;
+        }
+
+        /// <summary>
+        ///   Returns a new collection containing valid client certificates from the given X509Certificate2Collection
+        /// </summary>
+        internal static bool TryFindClientCertificate(this X509Certificate2Collection certificates,
+                                                      ISet<string> allowedIssuers,
+                                                      out X509Certificate2 clientCertificate,
+                                                      out X509Chain clientCertChain)
+        {
+            clientCertificate = null;
+            clientCertChain = null;
+            if (certificates == null)
+            {
+                return false;
+            }
+
+            DateTime now = DateTime.Now;
+            foreach(X509Certificate2 cert in certificates)
+            {
+                if (cert.HasPrivateKey && now > cert.NotBefore && now < cert.NotAfter)
+                {
+                    if (IsClientCertificate(cert))
+                    {
+                        X509Chain chain = new X509Chain();
+                        //TODO (#2204) set it to AllFlag when 2204 is fixed.
+                        chain.ChainPolicy.VerificationFlags = X509VerificationFlags.NoFlag;
+                        chain.ChainPolicy.RevocationFlag = X509RevocationFlag.ExcludeRoot;
+                        chain.ChainPolicy.ApplicationPolicy.Add(new Oid(ClientCertificateOid));
+                        if (!chain.Build(cert))
+                        {
+                            // TODO: uncomment 'continue' based on decision around 2204
+
+                            // Ideally, we would like to set VerificationFlags = AllFlags
+                            // But we are getting "work in progress" exception here when verificationFlags is anything but NoFlag
+                            // continue;
```

Comment:
I'm still confused by this comment.  I don't know why you want AllFlags, and don't know why you aren't continuing.